File: src/Cpp/11_Napisy_wprowadzenie/Zad10.cpp
#include <cassert>
#include <string>
#include <algorithm>
#include <vector>

//Otrzymujesz napis reprezuntajacy zdanie. 
//Znajdz najdluzsze oraz najkrotsze slowo w zdaniu. 
//Jesli wiecej niz jedno slowo ma ekstremalna dlugosc, 
//zwroc te, ktore wystepuje w zdaniu jako pierwsze. 
//Znaki interpunkcyjne nie sa liczone jako slowa. 

void wyczysc(std::string& napis){
	auto it = napis.begin();

	while (it != napis.end()) {
		if (ispunct(*it))
			napis.erase(it);
		else
			it++;
	}
}

std::string najdluzsze(std::string& napis) {
	std::string wynik;
	unsigned int dlugosc = 0;
	unsigned int pocz = 0; 
	unsigned int konc = 0;
	while ((konc = napis.find(' ', pocz)) != std::string::npos) {
		if (konc != pocz) {
			auto slowo = napis.substr(pocz, konc - pocz);
			wyczysc(slowo);

			if (slowo.size() > dlugosc) {
				wynik = slowo;
				dlugosc = slowo.size();

			}
		}
		pocz = konc + 1;
	}
	if (konc != pocz) {
		auto slowo = napis.substr(pocz);
		wyczysc(slowo);
		if (slowo.size() > dlugosc) {
			wynik = slowo;
			dlugosc = slowo.size();		
		}
	}

	return wynik;
}

std::string najkrotsze(std::string& napis) {
	std::string wynik;
	unsigned int dlugosc = napis.size();
	unsigned int pocz = 0; 
	unsigned int konc = 0;
	while ((konc = napis.find(' ', pocz)) != std::string::npos) {
		if (konc != pocz) {
			auto slowo = napis.substr(pocz, konc - pocz);
			wyczysc(slowo);

			if (slowo.size() < dlugosc) {
				wynik = slowo;
				dlugosc = slowo.size();

			}
		}
		pocz = konc + 1;
	}
	if (konc != pocz) {
		auto slowo = napis.substr(pocz);
		wyczysc(slowo);
		if (slowo.size() < dlugosc) {
			wynik = slowo;
			dlugosc = slowo.size();		
		}
	}

	return wynik;
}

void test1() {
	std::string napis = "Kaczka lubi wiosne.";
	std::string wynik = "Kaczka";
	assert(najdluzsze(napis) == wynik);
}

void test2() {
	std::string napis = "Kaczka lubi wiosne.";
	std::string wynik = "lubi";
	assert(najkrotsze(napis) == wynik);
}

int main() {
	test1();
	test2();
	return 0;
}

File: src/Python/11_Napisy_wprowadzenie/Zad10.py
import string

"""
Otrzymujesz napis reprezentujacy zdanie. Znajdz: 
a) Najdluzsze slowo w zdaniu.
b) Najkrotsze slowo w zdaniu. 

Jesli wiecej niz jedno slowo ma ekstremalna dlugosc, zwroc to 
slowo sposrod nich, ktore wystepuje w zdaniu jako pierwsze. 
Znaki interpunkcyjne nie sa liczone jako slowa.  
"""


def podziel_zdanie_na_slowa(zdanie):
    return zdanie.translate(str.maketrans("", "", string.punctuation)).split()


def najdluzsze_slowo(zdanie):
    slowa = podziel_zdanie_na_slowa(zdanie)
    return max(slowa, key=len)


def najkrotsze_slowo(zdanie):
    slowa = podziel_zdanie_na_slowa(zdanie)
    return min(slowa, key=len)


def test_najdluzsze_slowo():
    assert najdluzsze_slowo("Ala ma kota.") == "kota"


def test_najkrotsze_slowo():
    assert najkrotsze_slowo("Ala ma kota.") == "ma"


if __name__ == "__main__":

    test_najdluzsze_slowo()
    test_najkrotsze_slowo()
